persistence/docs/SAVEPOINTS_ROLLBACKS.md

```diff
@@ -0,0 +1,146 @@
+
+We either **succeed** across **all** data stores or we **fail** and we have to be able to **recover** from that failure like if **nothing has happened**.
+
+
+The following diagram illustrates the high-level flow:
+
+```mermaid
+flowchart TD
+    NewBlock[A Block was just committed\nor we are at Height 0] -->CreateSavepoint(Create SavePoint)
+    CreateSavepoint --> UpdatingState(We are creating\na new block.\nUpdating the State)
+    UpdatingState --> ShouldCommit
+    ShouldCommit{Should we commit?} --> |Yes| ReceivedOKFromDatastores{Have all\nthe datastores\ncommitted\nsuccessfully?}
+    ShouldCommit --> |No| NewBlock
+    ReceivedOKFromDatastores -->|Yes| Committed[Committed successfully] --> |Repeat with\nthe new Block| NewBlock
+    ReceivedOKFromDatastores -->|No, we have at least one failure| Rollback[Rollback] --> |Repeat with\nthe rolled back Block | NewBlock
+
+```
+
+## Definitions
+
+### Savepoints
+
+A savepoint (also referred as checkpoints/snapshots depending on the implementation) is either the beginning of database transaction (or distributed transaction) or some sort of artifact that was created right after a successful commit happened that allows to recreate a perfect copy of the state at the time it was created.
+
+### Rollbacks
+
+A rollback is the process of cleanly reverting the state of the node to a previous state that we have saved in a savepoint.
+
+### Minimum Viable Product
+
+After having examined the `Persistence` and `Utility` modules, I have identified the following areas that we can consider as part of the MVP, these could be used as individual Github issues/deliverables potentially:
+
+- [**State changes invalidation and rollback triggering**] We need some sort of shared and thread-safe reference that is available for us across the whole call-stack that we can use in the event of a failure to flag that we need to abort whatever we are doing and rollback. This could be achieved via the use of the [context](https://pkg.go.dev/context) package.
+
+- [**Ensure atomicity across data stores**] We need to make sure that we are using transactions correctly and that we are not accidentally committing state ahead of time in any of the data-stores.
+
+- [**Distributed commits across data-stores**] We need to implement a 2PC (two-phase commit) or 3PC (three-phase commit) protocol that should make sure that the state has been committed safely on **all** data stores before it is considered `valid`.
+
+    > TODO: **Discuss what is appropriate to do here**
+
+- [**Savepoints**] The simplest version could be the database transaction that can simply be discarded, basically the uncommitted transaction sits in memory until it's flushed to storage and rolling back to a savepoint would be as simple as discarding the non-pristine version of the state in memory.
```

Comment:
Just to make sure I understand, here's an example of what I envisage:

1. We have a committed height of 5 and we are at now, tentatively, at height 6, reaping the mempool, updating the trees, computing the root, computing the state hash
2. something (could be anything, including the fact that the state hash is not valid or non-consensus) goes wrong
3. the ephemeral changes are rolled back and we are back at the point **1** aka. the whole state is like it was at height 5 and we can try again

Am I missing something?

Reply:
Nope, we're on the same page.

- We use `postgres` for state change validation (can A send X POKT to B?)
- We use trees (k-v stores) to efficiently share the state everyone is agreeing on

- For postgers - rollback is easy because we keep reading/writing to an uncommited transaction and can just do `tx.rollback`.
- For KV stores, there is no `kv.rollback` without keeping track of all the modified keys

Reply:
💯